sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+        if sqlserver_adapter.APPROXIMATED_TYPES.get(old_type) == new_type:
+            new_col_dict["dataType"] = new_type = old_type
+
+        # Geometry type loses its extra type info when roundtripped through SQL Server.
+        if new_type == "geometry":
+            new_col_dict["geometryType"] = old_col_dict.get("geometryType")
+            new_col_dict["geometryCRS"] = old_col_dict.get("geometryCRS")
+
+        return new_type == old_type
+
+    def _remove_hidden_meta_diffs(self, dataset, ds_meta_items, wc_meta_items):
+        super()._remove_hidden_meta_diffs(dataset, ds_meta_items, wc_meta_items)
+
+        # Nowhere to put these in postgis WC
+        for key in self._UNSUPPORTED_META_ITEMS:
+            if key in ds_meta_items:
+                del ds_meta_items[key]
+
+        # Diffing CRS is not yet supported.
+        for key in list(ds_meta_items.keys()):
+            if key.startswith("crs/"):
+                del ds_meta_items[key]
+
+    def _is_meta_update_supported(self, dataset_version, meta_diff):
+        """
+        Returns True if the given meta-diff is supported *without* dropping and rewriting the table.
+        (Any meta change is supported if we drop and rewrite the table, but of course it is less efficient).
+        meta_diff - DeltaDiff object containing the meta changes.
+        """
+        # For now, just always drop and rewrite.
+        return not meta_diff
+
+
+class InstanceFunction(Function):
+    """
+    An instance function that compiles like this when applied to an element:
+    >>> element.function()
+    Unlike a normal sqlalchemy function which would compile as follows:
+    >>> function(element)
+    """
+
+    pass
+
+
+@compiles(InstanceFunction)
+def compile_instance_function(element, compiler, **kw):
+    return "(%s).%s()" % (element.clauses, element.name)
+
+
+class GeographyType(UserDefinedType):
+    """UserDefinedType so that V2 geometry is adapted to MS binary format."""
+
+    def __init__(self, crs_id):
+        self.crs_id = crs_id
+
+    def bind_processor(self, dialect):
+        # 1. Writing - Python layer - convert sno geometry to WKB
+        return lambda geom: geom.to_wkb()
+
+    def bind_expression(self, bindvalue):
+        # 2. Writing - SQL layer - wrap in call to STGeomFromWKB to convert WKB to MS binary.
+        return Function(
+            quoted_name("geography::STGeomFromWKB", False),
+            bindvalue,
+            self.crs_id,
+            type_=self,
+        )
+
+    def column_expression(self, col):
+        # 3. Reading - SQL layer - append with call to .STAsBinary() to convert MS binary to WKB.
+        return InstanceFunction("STAsBinary", col, type_=self)
+
+    def result_processor(self, dialect, coltype):
+        # 4. Reading - Python layer - convert WKB to sno geometry.
+        return lambda wkb: Geometry.from_wkb(wkb)
+
+
+class BaseDateOrTimeType(UserDefinedType):
```

Comment:
Neither ODBC/MSSQL/SpatialAlchemy have any handling for dates? Why do we want them as strings?

Reply:
Dates are put into the database as dates - that's fine, we declare the column type as "date" or similar and then any string we try to insert in there turns into a date - but we read them out as strings, using this UserDefinedType. This is because in Datasets V2, we store dates, times, etc as strings, so that's what we need to read when we want to do a diff / commit. The strings we store in Datasets V2 are ISO8601 formatted, which happily defines a format for dates, times, datetimes, and intervals.

This conversion could have happened at a different layer - eg we could have read dates from the database as Python datetimes, and then only convert them at a different Datasets V2 layer - but it doesn't, for a few reasons. 

- One is that Datasets V2 is our lingua franca, and I didn't see the point in defining another format for the layer between the database and datasets V2 - so things are read out of the database into V2 format immediately. This means they go straight into a format that can be easily serialised in message pack (strings are good) and skip more pythonic representations (datetimes).

- Another is that different working copies support different things and so doing the conversion immediately at the database layer straight back into V2 makes sense, since at that point you know the entire conversion you need to do. I could have decided to read PostGIS intervals as python timedeltas, and then converted them to something serialisable at another layer. But then SQL server doesn't have an interval type. So in that case, I would be reading a string from the database, converting it into a timedelta, then converting that back into a string, then serialising using message pack.

- Finally, having everything be immediately converted into our message-pack serializable V2 format, also means everything is JSON serialisable, and the diffs can easily be output to the user as text. It would be bad if we were comparing timedeltas and we decided they were different, but when we tried to show them to the user, the textual representation of them somehow ended up the same.

Maybe that was too much info

Reply:
More usefully, added a comment explaining that its only during reading that we read them as text

Reply:
Makes sense 👍